go-controller/pkg/util/multi_network.go

```diff
@@ -60,20 +73,346 @@ type BasicNetInfo interface {
+func ReconcileNetwork(to ReconcilableNetInfo, from NetInfo) error {
+	if from == nil || to == nil {
+		return fmt.Errorf("can't reconcile a nil network")
+	}
+	if !AreNetworksCompatible(to, from) {
+		return fmt.Errorf("can't reconcile from incompatible network")
+	}
+	reconcilable(to).reconcile(from)
+	return nil
+}
+
+func copyNetInfo(netInfo NetInfo) any {
+	switch t := netInfo.GetNetInfo().(type) {
+	case *DefaultNetInfo:
+		return t.copy()
+	case *secondaryNetInfo:
+		return t.copy()
+	default:
+		panic(fmt.Errorf("unrecognized type %T", t))
+	}
+}
+
+func reconcilable(netInfo NetInfo) ReconcilableNetInfo {
+	switch t := netInfo.GetNetInfo().(type) {
+	case *DefaultNetInfo:
+		return t
+	case *secondaryNetInfo:
+		return t
+	default:
+		panic(fmt.Errorf("unrecognized type %T", t))
+	}
+}
+
+// mutableNetInfo contains network information that can be changed
+type mutableNetInfo struct {
+	sync.RWMutex
+
+	nads                     sets.Set[string]
+	podNetworkAdvertisements map[string][]string
+	eipAdvertisements        map[string][]string
+
+	// information generated from previous fields, not used in comparisons
+
+	// namespaces from nads
+	namespaces sets.Set[string]
+}
+
+func mutable(netInfo NetInfo) *mutableNetInfo {
+	switch t := netInfo.GetNetInfo().(type) {
+	case *DefaultNetInfo:
+		return &t.mutableNetInfo
+	case *secondaryNetInfo:
+		return &t.mutableNetInfo
+	default:
+		panic(fmt.Errorf("unrecognized type %T", t))
+	}
+}
+
+func (l *mutableNetInfo) needsReconcile(r NetInfo) bool {
+	return !mutable(r).equals(l)
+}
+
+func (l *mutableNetInfo) reconcile(r NetInfo) {
+	l.copyFrom(mutable(r))
+}
+
+func (l *mutableNetInfo) equals(r *mutableNetInfo) bool {
+	l.RLock()
+	defer l.RUnlock()
+	r.RLock()
+	defer r.RUnlock()
+	return reflect.DeepEqual(l.nads, r.nads) &&
+		reflect.DeepEqual(l.podNetworkAdvertisements, r.podNetworkAdvertisements) &&
+		reflect.DeepEqual(l.eipAdvertisements, r.eipAdvertisements)
+}
+
+func (l *mutableNetInfo) copyFrom(r *mutableNetInfo) {
+	aux := mutableNetInfo{}
+	r.RLock()
+	aux.nads = r.nads.Clone()
+	aux.setPodNetworkAdvertisedOnVRFs(r.podNetworkAdvertisements)
+	aux.setEgressIPAdvertisedAtNodes(r.eipAdvertisements)
+	aux.namespaces = r.namespaces.Clone()
+	r.RUnlock()
+	l.Lock()
+	defer l.Unlock()
+	l.nads = aux.nads
+	l.podNetworkAdvertisements = aux.podNetworkAdvertisements
+	l.eipAdvertisements = aux.eipAdvertisements
+	l.namespaces = aux.namespaces
+}
+
+func (nInfo *mutableNetInfo) SetPodNetworkAdvertisedVRFs(podAdvertisements map[string][]string) {
+	nInfo.Lock()
+	defer nInfo.Unlock()
+	nInfo.setPodNetworkAdvertisedOnVRFs(podAdvertisements)
+}
+
+func (nInfo *mutableNetInfo) setPodNetworkAdvertisedOnVRFs(podAdvertisements map[string][]string) {
+	nInfo.podNetworkAdvertisements = make(map[string][]string, len(podAdvertisements))
+	for node, vrfs := range podAdvertisements {
+		nInfo.podNetworkAdvertisements[node] = sets.List(sets.New(vrfs...))
```

Comment:
Yeah, so that many different controllers can access and read this info, just has we did with `GetNADs()`. Not update it, `NetInfo` is what we usually move around and does not have update methods any more.